### DISCUSSION TITLE
Artisan, Linux Mint (Ubuntu derivative), Mastech 6514 - "Serial Exception: Unable to open serial port" error

### DISCUSSION THREAD

**Discussion by Alaskahawk7** [artisan-roaster-scope/artisan#1355] [Q&A] [answered]
I downloaded and installed Artisan on a laptop running Linux Mint (uses Ubuntu 22.04.1).  I followed the instructions in the artisan/wiki/installation where it says in Step-4:

**_"Step 4: Configure Artisan for your setup
You need to tell Artisan which machine or devices you attached. Startup Artisan and select your roasting machine (menu Config >> Machines) or open the Device Assignment dialog (menu Config >> Device) and configure your device here. The serial settings for meters are already configured by Artisan automatically when you select a device for the first time (or when you change devices). The only setting that is not configured is the serial port number being used. To find out your serial port, connect your device/meter and select the correct comm port from the ports popup menu. You can test to see if you have the correct comm port by clicking the green button "ON" on the top of the main window. If you see the two temperatures from the meter come up on the LCDs, you have completed all the configuration steps."_**

In setting up the Mastech 6514 device it prompts for a Comm Port.  It looks like the default Comm Port is /dev/ttyUSB0 and I have tried that.  I get the error message "Serial Exception: Unable to open serial port."  I had already followed the artisan wiki instructions for Linux and uninstalled the BRLTTY package that reportedly creates a conflict for Artisan.  There is another Comm Port option in the dropdown box in Artisan in the Ports setting, "CP2102 USB to UART Bridge Controller." But I get the same error code when I select that one too.  Finally I looked under the /dev/usb/001/004 path and found that when I plug in the Mastech device it seems to show up as 004 under the 001 directory.  So I tried using that path in the Comm Port settings drop down box.  But I get the same error.  I don't understand how to fix this problem.  I'd sure appreciate any help you can give me.

**Comment by darrenaddy** [comment]
One step often missed on the Mastech 6514 is that you *have* to press and
hold the *PC* Link button until it beeps for the connection to go through.
Otherwise, regarding  Comm Port option in the dropdown box in Artisan in
the Ports setting, "CP2102 USB to UART Bridge Controller."
ChatGPT says this:

The Silicon Labs CP210x USB to UART bridge driver, as mentioned in the
documentation, is indeed a part of the Linux kernel from version 2.6.12
onwards. This means that the driver is already integrated into the kernel,
and in most cases, you don't need to manually install it. However, to use
it or get a USB device recognized in your Linux system, there are a few
things you might need to check or do:

   1.

   *Kernel Version*: Ensure your Linux system is running on kernel version
   2.6.12 or later. You can check your kernel version by running the
command uname
   -r in the terminal.
   2.

   *Module Loading*: Even though the driver is part of the kernel, it might
   not be loaded automatically. You can check if the module is loaded by
   running lsmod | grep cp210x. If it's not loaded, you can load it
   manually using sudo modprobe cp210x.
   3.

   *Device Recognition*: When you connect your USB device, it should be
   automatically recognized by the system. You can check this by looking at
   the output of dmesg | tail after plugging in the device. This command
   shows the kernel messages and should indicate whether the device was
   recognized and which tty device it is associated with (e.g., /dev/ttyUSB0).
   4.

   *Permissions*: Sometimes, the device may be recognized but not
   accessible due to permission issues. Make sure your user has the necessary
   permissions to access the device. This can often be resolved by adding your
   user to the dialout group with sudo usermod -a -G dialout $USER, and
   then logging out and back in.
   5.

   *Udev Rules*: In some cases, you might need to create or modify udev
   rules to set proper permissions or perform certain actions when the device
   is connected. This is more advanced and typically only necessary for
   specific use cases.
   6.

   *Software Dependencies*: Ensure that any software you intend to use with
   the USB device is compatible and properly configured to communicate with
   the device.
   7.

   *Troubleshooting*: If the device is not recognized, check the cables and
   ports, and try connecting to a different port or computer to rule out
   hardware issues.

In summary, in most scenarios, the driver should work out of the box with
minimal intervention. If you encounter issues, the steps above can help
troubleshoot and resolve common problems.
</end ChatGPT4+>
Hope this helps!

**Reply by Alaskahawk7** [reply, marked as answer]
I tried these and that option #4 (adding the user, me, to the dialout group) did remove the original error message.   That was the issue.  Everything is working great now.  I (at first) thought I had a problem due to the data not being read by my laptop.  However I had not held the "PC Link" button down long enough for the "USB" to appear on the screen of the Mastech.  Once I did that though, the data was flowing to the Artisan software and being read.  The initial error though was fixed by adding my username to the dialout group.  Thanks for the help, I really appreciate it.

**Reply by sinapythoning** [reply]
Hi, I hope all is well with you. I just got a MS6514 and am trying to connect it to my roaster and I think I have a similar problem but I couldn’t figure it out on my own. Do you have time for a zoom call to help me out? You can reach me at these emails. sbasiri@iu.edu/sinabasiri2000@gmail.com

**Comment by darrenaddy** [comment]
When I put that error into Google, I get this from a fellow Mastech 6514 on
Linux user:

This is because the serial port configuration (Config -> Serial Port) in
Artisan is not set properly.
I use Artisan v1.1 (last 32bit version) under Linux and these are my
settings:

---
Baud rate 9600
Byte size 8
Parity N
Stop bits 1
Time out 1.0
- source thread:
https://www.home-barista.com/roasting/need-help-with-mastech-ms6514-artisan-t76409.html

This thread also mentions getting the same error and the fix:
https://homeroasters.org/forum/viewthread.php?thread_id=5169

Hope this helps.

**Comment by darrenaddy** [comment]
At the risk of repeating myself, I found the same error earlier and it was
simply that the user was not doing the following on his Mastech 6514:

> Did you long-push the “PC-Link” button to turn
> communication on as advised in the manual?

That worked!

Source: https://github.com/lammy/artisan/issues/79

Hope this helps.
